core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.Tasks;
+
+import static org.apache.iceberg.TableProperties.COMMIT_MAX_RETRY_WAIT_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_MAX_RETRY_WAIT_MS_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_MIN_RETRY_WAIT_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_MIN_RETRY_WAIT_MS_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_NUM_RETRIES;
+import static org.apache.iceberg.TableProperties.COMMIT_NUM_RETRIES_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT;
+
+class UpdateSnapshotReferencesOperation implements PendingUpdate<Map<String, SnapshotRef>> {
+
+  private final TableOperations ops;
+  private TableMetadata base;
+  private final Map<String, SnapshotRef> refsToUpdate;
+  private final Set<String> refsToRemove;
+
+  UpdateSnapshotReferencesOperation(TableOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+    this.refsToUpdate = Maps.newHashMap();
+    this.refsToRemove = Sets.newHashSet();
+  }
+
+  @Override
+  public Map<String, SnapshotRef> apply() {
+    this.base = ops.refresh();
+    Map<String, SnapshotRef> refs = Maps.newHashMap();
+    TableMetadata.Builder updatedBuilder = TableMetadata.buildFrom(base);
+    for (String refName : refsToRemove) {
+      SnapshotRef ref = base.ref(refName);
+      ValidationException.check(ref != null, "Cannot remove nonexistent snapshot ref %s", refName);
+      if (ref.isBranch()) {
+        updatedBuilder.removeBranch(refName);
+      }
+      else {
+        updatedBuilder.removeTag(refName);
+      }
+    }
+    for (Map.Entry<String, SnapshotRef> refEntry : refsToUpdate.entrySet()) {
+      String name = refEntry.getKey();
+      SnapshotRef snapshotRef = refEntry.getValue();
+      updatedBuilder.createSnapshotRef(name, snapshotRef);
+    }
+    return refs;
+  }
+
+  @Override
+  public void commit() {
+    Tasks.foreach(ops)
+        .retry(base.propertyAsInt(COMMIT_NUM_RETRIES, COMMIT_NUM_RETRIES_DEFAULT))
+        .exponentialBackoff(
+            base.propertyAsInt(COMMIT_MIN_RETRY_WAIT_MS, COMMIT_MIN_RETRY_WAIT_MS_DEFAULT),
+            base.propertyAsInt(COMMIT_MAX_RETRY_WAIT_MS, COMMIT_MAX_RETRY_WAIT_MS_DEFAULT),
+            base.propertyAsInt(COMMIT_TOTAL_RETRY_TIME_MS, COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT),
+            2.0 /* exponential */)
+        .onlyRetryOn(CommitFailedException.class)
+        .run(taskOps -> {
+          Map<String, SnapshotRef> refs = apply();
+          TableMetadata.Builder updatedMetadata = TableMetadata.buildFrom(base);
+          updatedMetadata.
+          taskOps.commit(base, updated);
+        });
+  }
+
+  public UpdateSnapshotReferencesOperation createBranch(String name, long snapshotId) {
```

Comment:
I think these methods don't quite get the behavior correct. In the table update APIs, changes should be applied in sequence so that each one updates the table as though it were part of its own commit. For example, in the `UpdateSchema` API, two calls to `moveFirst` will move one column, then the second. The two operations don't conflict.

Here, calling `createBranch` followed by `setMinSnapshotsToKeep` on the same branch will result in an error because this adds the new ref to `refsToUpdate` and the other finds a ref in `table.refs()`.

I think this class should start by copying the table refs, then make changes to that set of refs. In `apply`, this can diff the original set of refs against the current table refs to make specific changes to `TableMetadata`. For table property updates, we do that diff in `TableMetadata` and I think it would be reasonable to do the same for this if you want.

Reply:
@rdblue I was still working on changes and iteratively pushing to my branch (sorry for the confusion),  I think this was reviewed when it wasn't yet complete on my end. 

For supporting transaction.updateSnapshotReferencesOperation().createBranch(...).setMinSnapshotsToKeep() to avoid the error, in my latest change when we look for refs in setMinSnapshotsToKeep we always search for the "current" ref. Where current is either the ref in the refsToUpdate (which if someone did a createBranch, it would be in there) or we then fall back to the ref. validateAndGetLatestRef is the helper method to accomplish this. So effectively it's the same as operating on a copy of the refs.

Currently, in apply we do the diff to keep track of the metadata changes.

Reply:
Okay, as long as this achieves the expected behavior it should be okay. It does seem like copying the refs into a new map and only working with that map is going to be the most straightfoward approach though. Falling back means you have to remember to do it.

Reply:
Yeah, I think when I enumerate the different cases explicitly in tests we will get better confidence on if this is the right way. I will also look into just using a copy of the refs.

Reply:
Updated to just use a copy of the refs